### PR TITLE
Refactor confirm email page to use common user flow function

### PR DESCRIPTION
**Pull request by kentsanggds**
## What

Refactor the test to use the common user flow function

## How to review

execute run-tests

### CODE REVIEW

File: tests/functional/test_confirm_email_address.py
Comment (andyhd) [inline tests/functional/test_confirm_email_address.py]: Don't have multiple lines of imports from the same module, instead do this:
```python
from tests.function.user_flows import (
    on_email_confirm...,
    yes_selected,
    ...
)
```

Comment (andyhd) [reply tests/functional/test_confirm_email_address.py]: Are these imports expected to be used across a number of test modules? I don't see any need for them to be in a separate file from the tests that use them if not.

File: tests/functional/test_confirm_email_address.py
Comment (andyhd) [inline tests/functional/test_confirm_email_address.py]: This would be more self-documenting as two separate function calls, eg:
```python
set_email_known(True)
submit_email(email_address)
```

Comment (andyhd) [reply tests/functional/test_confirm_email_address.py]: Also, since these are essentially fixtures, you could write them like this to avoid passing `browser` around:
```python
@pytest.fixture
def set_email_known(browser):
    def do_set_email_known(known=True):
        value = 'yes' if known else 'no'
        browser.choose('email_known', value)
    return do_set_email_known
```
and then in the test
```python
def it_goes_to_department_confirm_when_continue_clicked(browser, set_email_known, submit_email):
    email_address = 'some.one@...'
    set_email_known()
    submit_email(email_address)
    ...
```

File: tests/functional/test_confirm_email_address.py
Comment (andyhd) [inline tests/functional/test_confirm_email_address.py]: Since all the tests in this class will use the `live_server` fixture, you can move it out of the method signature and put it in a class decorator like this:
```python
@pytest.mark.usefixtures('live_server')
class When_on_confirm_email_with_yes_selected_and_valid_email(object):
    def it_goes_to_department_confirm_when_continue_clicked(self, browser):
        ...
```

**Review by andyhd — CHANGES_REQUESTED** [review]

File: tests/functional/test_confirm_email_address.py
Comment (kentsanggds) [reply tests/functional/test_confirm_email_address.py]: ok, and yes they will be used in different functional tests

File: tests/functional/test_confirm_email_address.py
Comment (kentsanggds) [reply tests/functional/test_confirm_email_address.py]: ok, will do that

File: tests/functional/test_confirm_email_address.py
Comment (kentsanggds) [reply tests/functional/test_confirm_email_address.py]: ok

File: tests/conftest.py
Comment (andyhd) [inline tests/conftest.py:3]: 3rd party modules should be imported before local modules. The order of imports should be:
* core modules (eg: `os`, `urllib`)
* 3rd party modules (eg: `flask`, `responses`)
* local modules (eg: `app.main`, `tests.oidc_testbed`)

with a single blank line between each set and two blank lines after the last import

File: tests/conftest.py
Comment (andyhd) [inline tests/conftest.py]: This line doesn't need to be split

File: tests/functional/test_confirm_email_address.py
Comment (andyhd) [inline tests/functional/test_confirm_email_address.py]: Just a suggestion, feel free to ignore.

Since this class sets up the test context as "When on confirm email (page), with yes selected and valid email (submitted)", you could use a `setup_method` to set up that context programmatically, eg:
```python
@pytest.mark.usefixtures('live_server')
class When_on_confirm_email_with_yes_selected_and_valid_email(object):
    email_address = 'some.one@digital.cabinet-office.gov.uk'
    department = 'Government Digital Service'

    def setup_method(self):
        set_email_known(True)
        set_email_address(self.email_address)

    def it_goes_to_department_confirm_when_continue_clicked(self, browser):
        browser.find_by_css('form button').click()
        assert browser.url == url_for('main.confirm_dept', _external=True)
        ...
```
This would make this test class pretty close to a BDD style given/when structure

File: tests/functional/test_confirm_email_address.py
Comment (andyhd) [inline tests/functional/test_confirm_email_address.py]: This line doesn't need to be broken

**Review by andyhd — CHANGES_REQUESTED** [review]

File: tests/conftest.py
Comment (kentsanggds) [reply tests/conftest.py:3]: ok, initially reordered them alphabetically but I guess need to group them first

File: tests/conftest.py
Comment (kentsanggds) [reply tests/conftest.py]: ok

File: tests/functional/test_confirm_email_address.py
Comment (kentsanggds) [reply tests/functional/test_confirm_email_address.py]: Think I'll just leave the code as it is set up so that I can move on, may try this at a later stage though

File: tests/functional/conftest.py
Comment (andyhd) [inline tests/functional/conftest.py]: `email_address` is not used in this fixture, because it is shadowed by the `email_address` parameter to `do_submit_known_email_address`

Comment (andyhd) [reply tests/functional/conftest.py]: In fact, you are not using this fixture anywhere - is it still needed?

File: tests/functional/test_confirm_email_address.py
Comment (andyhd) [inline tests/functional/test_confirm_email_address.py]: This might be more readable as:
```python
assert browser.find_by_name('email_address').visible is False
```
What do you think?

**Review by andyhd — APPROVED** [review]

**Comment by andyhd** [comment]
Waiting until after assessment to merge

File: tests/functional/conftest.py
Comment (kentsanggds) [reply tests/functional/conftest.py]: ok, will be using this fixture for other tests, can take it out and add it when needed

File: tests/functional/test_confirm_email_address.py
Comment (kentsanggds) [reply tests/functional/test_confirm_email_address.py]: ok

**Pull request merged**
